pkg/controller/constrainttemplate/stats_reporter.go

```diff
@@ -0,0 +1,143 @@
+	"go.opencensus.io/tag"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	ctCount        = "total_constraint_templates"
+	ingestCount    = "constraint_template_ingestion_count"
+	ingestDuration = "constraint_template_ingestion_duration_seconds"
+)
+
+var (
+	ctCountM        = stats.Int64(ctCount, "Total number of constraint templates", stats.UnitDimensionless)
+	ingestDurationM = stats.Float64(ingestDuration, "Total time to ingest a constraint template in seconds", stats.UnitSeconds)
+
+	statusKey = tag.MustNewKey("status")
+
+	views = []*view.View{
+		{
+			Name:        ctCount,
+			Measure:     ctCountM,
+			Description: "Total number of constraint templates",
+			Aggregation: view.LastValue(),
+			TagKeys:     []tag.Key{statusKey},
+		},
+		{
+			Name:        ingestCount,
+			Measure:     ingestDurationM,
+			Description: "Total number of constraint template ingestion actions",
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{statusKey},
+		},
+		{
+			Name:        ingestDuration,
+			Measure:     ingestDurationM,
+			Description: "Total time to ingest a constraint template in seconds",
+			Aggregation: view.Distribution(0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1, 0.2, 0.3, 0.4, 0.5, 1, 2, 3, 4, 5),
+			TagKeys:     []tag.Key{statusKey},
+		},
+	}
+)
+
+type ctStatus string
+
+const (
+	statusActive = ctStatus("active")
+	statusError  = ctStatus("error")
+)
+
+func init() {
+	if err := register(); err != nil {
+		panic(err)
+	}
+}
+
+func register() error {
+	return view.Register(views...)
+}
+
+func reset() error {
+	view.Unregister(views...)
+	return register()
+}
+
+func (r *reporter) reportCtCount(status ctStatus, count int64) error {
+	ctx, err := tag.New(
+		r.ctx,
+		tag.Insert(statusKey, string(status)),
+	)
+	if err != nil {
+		return err
+	}
+	return metrics.Record(ctx, ctCountM.M(count))
+}
+
+func (r *reporter) reportIngestDuration(status ctStatus, d time.Duration) error {
+	ctx, err := tag.New(
+		r.ctx,
+		tag.Insert(statusKey, string(status)),
+	)
+	if err != nil {
+		return err
+	}
+
+	return metrics.Record(ctx, ingestDurationM.M(d.Seconds()))
+}
+
+// newStatsReporter creates a reporter for watch metrics
+func newStatsReporter() (*reporter, error) {
+	ctx, err := tag.New(
+		context.TODO(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	reg := &ctRegistry{cache: make(map[types.NamespacedName]ctStatus)}
+	return &reporter{ctx: ctx, registry: reg}, nil
+}
+
+type reporter struct {
+	ctx      context.Context
+	registry *ctRegistry
+}
+
+type ctRegistry struct {
+	cache map[types.NamespacedName]ctStatus
+}
+
+func (r *ctRegistry) add(key types.NamespacedName, status ctStatus) bool {
+	v, ok := r.cache[key]
+	if ok && v == status {
+		return false
+	}
+	r.cache[key] = status
+	return true
+}
+
+func (r *ctRegistry) remove(key types.NamespacedName) {
+	delete(r.cache, key)
+}
+
+func (r *ctRegistry) report(metrics *reporter) {
+	totals := map[ctStatus]int64{
+		statusError:  0,
+		statusActive: 0,
```

Comment:
Nit: no need to initialize these values. They are the defaults anyways and this block of code would need to be maintained as new status values get added.

Reply:
We need some source-of-truth for all observable values so we can handle when the # of constraints with a given count go to zero.

Example: there was one constraint with an error status, now there are none, because that means statusError will disappear from the map entirely, the gauge would never get updated.

The other option is to cache a list of all observed statuses.